newsroom/history_async.py

```diff
@@ -0,0 +1,178 @@
+                "versioncreated": now,
+                "user": user.id,
+                "company": user.company,
+                "item": item["_id"],
+                "version": str(item.get("version", item.get("_current_version"))),
+                "section": section,
+            }
+
+        for doc in docs:
+            try:
+                await super().create([await transform(doc)])
+            except (werkzeug.exceptions.Conflict, pymongo.errors.BulkWriteError):
+                continue
+
+    async def create_history_record(
+        self, items: List[Dict[str, Any]], action: str, user: UserResourceModel, section: str
+    ):
+        await self.create(items, action, user, section)
+
+    async def query_items(self, query: Dict[str, Any]):
+        if query["from"] >= 1000:
+            # https://www.elastic.co/guide/en/elasticsearch/guide/current/pagination.html#pagination
+            return abort(400)
+
+        req = ParsedRequest()
+        req.args = {"source": json.dumps(query)}
+        return await super().get(req, None)
+
+    async def fetch_history(self, query: Dict[str, Any], all: bool = False):
+        results = await self.query_items(query)
+        docs = results.docs
+        if all:
+            while results.hits["hits"]["total"]["value"] > len(docs):
+                query["from"] = len(docs)
+                results = await self.query_items(query)
+                docs.extend(results.docs)
+
+        return {"_items": docs, "hits": results.hits}
+
+
+async def get_history_users(
+    item_ids: List[Union[ObjectId, str]],
+    active_user_ids: List[Union[ObjectId, str]],
+    active_company_ids: List[Union[ObjectId, str]],
+    section: str,
+    action: str,
+) -> List[str]:
+    source = {
+        "query": {
+            "bool": {
+                "filter": [
+                    {
+                        "bool": {
+                            "should": [
+                                {"terms": {"company": [str(a) for a in active_company_ids]}},
+                                {"bool": {"must_not": [{"exists": {"field": "company"}}]}},
+                            ],
+                            "minimum_should_match": 1,
+                        },
+                    },
+                    {"terms": {"item": [str(i) for i in item_ids]}},
+                    {"term": {"section": section}},
+                    {"term": {"action": action}},
+                ]
+            }
+        },
+        "size": 25,
+        "from": 0,
+    }
+
+    # Get the results
+    histories = await HistoryService().fetch_history(source, all=True)
+    histories_items = histories._items or []
+    # Filter out the users
+    user_ids = [str(uid) for uid in active_user_ids]
+    return [str(h.user) for h in histories_items if h.user in user_ids]
+
+
+history_resource_config = ResourceConfig(
+    name="history",
+    data_class=HistoryResourceModel,
+    service=HistoryService,
+    mongo=MongoResourceConfig(
```

Comment:
You need to [enable Elasticsearch](https://github.com/superdesk/superdesk-core/blob/async/tests/core/modules/users/resources.py#L36) for this resource

Reply:
yep